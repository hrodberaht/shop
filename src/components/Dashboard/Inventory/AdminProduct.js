import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ConnectedAddProductForm from './AddProductForm';
import { removeProduct, updateProduct } from '../../../store/products/actionCreators';
import { getAuthToken } from '../../../store/auth/selectors';

export class AdminProduct extends Component {
  state = {
    toggleEdit: false,
  };

  handleEditClick = () => {
    this.setState(state => ({
      toggleEdit: !state.toggleEdit,
    }));
  };

  handleRemoveClick = (id) => {
    const { removeProd, token } = this.props;
    removeProd(id, token);
  };

  submit = (values) => {
    const { updateProductInDB, token } = this.props;
    updateProductInDB(values, token);
  };

  render() {
    const {
      product,
      product: {
        id, name, type, price, inStock, remove,
      },
    } = this.props;
    return (
      <React.Fragment>
        {this.state.toggleEdit ? (
          <td colSpan="6">
            <ConnectedAddProductForm product={product} onSubmit={this.submit} />
            <button
              id="closeButton"
              className="btn btn-primary"
              type="button"
              onClick={this.handleEditClick}
            >
              Close
            </button>
          </td>
        ) : (
          <React.Fragment>
            <td>{name}</td>
            <td>{type}</td>
            <td>{+price}</td>
            <td>{+inStock}</td>
            <td>
              <button
                id="editButton"
                className="btn btn-primary"
                type="button"
                onClick={this.handleEditClick}
              >
                Edit
              </button>
            </td>
            <td>
              <button
                id="removeButton"
                className="btn btn-danger"
                type="button"
                onClick={() => this.handleRemoveClick(id)}
                disabled={remove}
              >
                X
              </button>
            </td>
          </React.Fragment>
        )}
      </React.Fragment>
    );
  }
}
const mapStateToProps = state => ({
  token: getAuthToken(state),
});

export default connect(
  mapStateToProps,
  {
    removeProd: removeProduct,
    updateProductInDB: updateProduct,
  },
)(AdminProduct);

AdminProduct.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    inStock: PropTypes.string.isRequired,
  }).isRequired,
  removeProd: PropTypes.func.isRequired,
  token: PropTypes.string.isRequired,
  updateProductInDB: PropTypes.func.isRequired,
};
