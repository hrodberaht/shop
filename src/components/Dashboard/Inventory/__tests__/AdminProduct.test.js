import { AdminProduct } from '../AdminProduct';

const defaultProps = {
  product: {
    id: 'product-1',
    type: 'Printer',
    name: 'ZX3',
    price: 400,
    inStock: 30,
    remove: false,
  },
  removeProd: jest.fn(),
  token: '1234',
  updateProductInDB: jest.fn(),
  undeletedProduct: jest.fn(),
};

const setup = buildSetup(AdminProduct, defaultProps);

describe('<AdminProduct />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('isEditVisable and isModalVisable in state should be false', () => {
    const { wrapper } = setup();
    expect(wrapper.state().isEditVisable).toBe(false);
    expect(wrapper.state().isModalVisable).toBe(false);
  });

  it('should change state after click button edit and show form', () => {
    const { wrapper } = setup();
    wrapper.find('#editButton').simulate('click');
    expect(wrapper.state().isEditVisable).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should show undeleteButton when product is remove', () => {
    const { wrapper } = setup();
    wrapper.setProps({
      product: {
        id: 'product-1',
        type: 'Printer',
        name: 'ZX3',
        price: 400,
        inStock: 30,
        remove: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it('should change state after click button close and hide form', () => {
    const { wrapper } = setup();
    wrapper.find('#editButton').simulate('click');
    wrapper.find('#closeButton').simulate('click');

    expect(wrapper.state().isEditVisable).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });
  it('should call remove after call handleConfirmClick ', () => {
    const { wrapper } = setup();

    const {
      handleConfirmClick,
      props: {
        removeProd,
        token,
        product: { id },
      },
    } = wrapper.instance();
    wrapper.find('#removeButton').simulate('click');
    handleConfirmClick();

    expect(removeProd).toHaveBeenCalledWith(id, token);
  });

  it('state isVisableModal should change to true after click remove', () => {
    const { wrapper } = setup();
    wrapper.find('#removeButton').simulate('click');

    expect(wrapper.state().isModalVisable).toBe(true);
  });

  it('should call updateProductinDB after call submit ', () => {
    const { wrapper } = setup();
    wrapper.instance().submit();

    expect(wrapper.instance().props.updateProductInDB).toHaveBeenCalled();
  });

  it('should call undeletedProduct after click undeletedProductButton', () => {
    const {
      wrapper,
      props: {
        product: { id },
      },
    } = setup();
    wrapper.setProps({
      product: {
        id: 'product-1',
        type: 'Printer',
        name: 'ZX3',
        price: 400,
        inStock: 30,
        remove: true,
      },
    });
    wrapper.find('#undeleteProductButton').simulate('click');
    expect(wrapper.instance().props.undeletedProduct).toHaveBeenCalledWith(id);
  });
});
