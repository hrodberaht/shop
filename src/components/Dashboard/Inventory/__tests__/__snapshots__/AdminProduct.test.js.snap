// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminProduct /> should change state after click button close and hide form 1`] = `
<Fragment>
  <td>
    ZX3
  </td>
  <td>
    <img
      alt="Printer"
    />
  </td>
  <td>
    Printer
  </td>
  <td>
    400
  </td>
  <td>
    30
  </td>
  <td>
    <button
      className="btn btn-primary"
      id="editButton"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </td>
  <td>
    <button
      className="btn btn-danger"
      id="removeButton"
      onClick={[Function]}
      type="button"
    >
      X
    </button>
  </td>
</Fragment>
`;

exports[`<AdminProduct /> should change state after click button edit and show form 1`] = `
<Fragment>
  <td
    className="edit-form"
    colSpan="6"
  >
    <Connect(ReduxForm)
      onSubmit={[Function]}
      product={
        Object {
          "id": "product-1",
          "inStock": "30",
          "name": "ZX3",
          "price": "400",
          "type": "Printer",
        }
      }
    />
    <button
      className="btn btn-danger"
      id="closeButton"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </td>
</Fragment>
`;

exports[`<AdminProduct /> should render without crash 1`] = `
<Fragment>
  <td>
    ZX3
  </td>
  <td>
    <img
      alt="Printer"
    />
  </td>
  <td>
    Printer
  </td>
  <td>
    400
  </td>
  <td>
    30
  </td>
  <td>
    <button
      className="btn btn-primary"
      id="editButton"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </td>
  <td>
    <button
      className="btn btn-danger"
      id="removeButton"
      onClick={[Function]}
      type="button"
    >
      X
    </button>
  </td>
</Fragment>
`;
